util/iptables: check for and use new iptables-restore 'wait' argument

dcbw opened the pull request:
iptables-restore did not previously perform any locking, meaning that
when callers (like kube-proxy) asked iptables-restore to write large
numbers of rules, the iptables-restore process might run in parallel
with other 'iptables' invocations in kubelet (hostports), docker,
and other software.  This causes errors like:

"CNI request failed with status 400: 'Failed to ensure that nat chain
POSTROUTING jumps to MASQUERADE: error checking rule: exit status 4:
iptables: Resource temporarily  unavailable."

or from Docker:

"Failed to allocate and map port 1095-1095: iptables failed:
iptables --wait -t nat -A DOCKER -p tcp -d 0/0 --dport 1095
-j DNAT --to-destination 10.1.0.2:1095 ! -i lbr0: iptables:
Resource temporarily unavailable.\n (exit status 4)"

iptables-restore "wait" functionality was added in iptables git
commit 999eaa241212d3952ddff39a99d0d55a74e3639e which
is not yet in a release.

See also https://bugzilla.redhat.com/show_bug.cgi?id=1417234

@eparis @knobunc @kubernetes/rh-networking @kubernetes/sig-network-misc @freehan @thockin @brendandburns

dcbw committed:
util/iptables: check for and use new iptables-restore 'wait' argument

iptables-restore did not previously perform any locking, meaning that
when callers (like kube-proxy) asked iptables-restore to write large
numbers of rules, the iptables-restore process might run in parallel
with other 'iptables' invocations in kubelet (hostports), docker,
and other software.  This causes errors like:

"CNI request failed with status 400: 'Failed to ensure that nat chain
POSTROUTING jumps to MASQUERADE: error checking rule: exit status 4:
iptables: Resource temporarily  unavailable."

or from Docker

"Failed to allocate and map port 1095-1095: iptables failed:
iptables --wait -t nat -A DOCKER -p tcp -d 0/0 --dport 1095
-j DNAT --to-destination 10.1.0.2:1095 ! -i lbr0: iptables:
Resource temporarily unavailable.\n (exit status 4)"

iptables-restore "wait" functionality was added in iptables git
commit 999eaa241212d3952ddff39a99d0d55a74e3639e but is NOT YET
in a released version of iptables.

See also https://bugzilla.redhat.com/show_bug.cgi?id=1417234

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/43575)
<!-- Reviewable:end -->

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -503,6 +518,18 @@ func getIPTablesWaitFlag(vstring string) []string {
+// Checks if iptables-restore has a "wait" flag
+func getIPTablesRestoreWaitFlag(exec utilexec.Interface) restoreWaitFlag {
+	glog.V(4).Infof("checking if iptables-restore can wait")
+	bytes, err := exec.Command(cmdIPTablesRestore, "--help").CombinedOutput()
```
Can't we use a min version like we do for the other wait flags?

dcbw replied:
@thockin it's not in a released version yet, and we're sure as hell gonna backport to RHEL

dcbw replied:
by which I mean both this PR to openshift, and the iptables patches

thockin replied:
Can you comment HERE that we do this instead of checking version because ...

cmluciano replied:
I thought the wait interval made it into 1.6.1

dcbw replied:
@cmluciano no, 1.6.1 was in January 2017, while the wait patches landed in March.  So I assume 1.6.2 and later.

thockin commented:
Can you comment that, at least.

On Thu, Mar 23, 2017 at 10:45 AM, Dan Williams <notifications@github.com>
wrote:

> *@dcbw* commented on this pull request.
> ------------------------------
>
> In pkg/util/iptables/iptables.go
> <https://github.com/kubernetes/kubernetes/pull/43575#discussion_r107737230>
> :
>
> > @@ -503,6 +518,18 @@ func getIPTablesWaitFlag(vstring string) []string {
>  	}
>  }
>
> +// Checks if iptables-restore has a "wait" flag
> +func getIPTablesRestoreWaitFlag(exec utilexec.Interface) restoreWaitFlag {
> +	glog.V(4).Infof("checking if iptables-restore can wait")
> +	bytes, err := exec.Command(cmdIPTablesRestore, "--help").CombinedOutput()
>
> by which I mean both this PR to openshift, and the iptables patches
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/43575#discussion_r107737230>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVEQfieOfDkFVffWVF9PQusrueFoqks5roq-ugaJpZM4MnEGZ>
> .
>

dcbw commented:
@thockin next question; should we still use a 2s wait interval?  I feel like that's too short for iptables-restore, given we've seen some > 20s invocations with 1000s of services/endpoints and our apparently new-enough-that-it-takes-longer kernels.

But kube-proxy is likely the only thing running restore with large rulesets, and that's already serialized, so we'd only be waiting for other users running 'iptables' not other iptables-restore processes.  So perhaps 2s is fine.  What do you think?

dcbw commented:
> Can you comment that, at least.

@thockin added to commit message.

thockin commented:
iptables -w2 flag is not "wait 2 seconds" but "v2 of the wait flag".  Is
this new flag different?

On Thu, Mar 23, 2017 at 10:55 AM, Dan Williams <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> next question; should we still use
> a 2s wait interval? I feel like that's too short for iptables-restore,
> given we've seen some > 20s invocations with 1000s of services/endpoints
> and our apparently new-enough-that-it-takes-longer kernels.
>
> But kube-proxy is likely the only thing running restore with large
> rulesets, and that's already serialized, so we'd only be waiting for other
> users running 'iptables' not other iptables-restore processes. So perhaps
> 2s is fine. What do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/43575#issuecomment-288806971>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVCdMQNF0i3eCPyBBTDQt0bdLMeVAks5rorIRgaJpZM4MnEGZ>
> .
>

dcbw commented:
> iptables -w2 flag is not "wait 2 seconds" but "v2 of the wait flag". Is this new flag different?

@thockin no, it's the actual option value in seconds.  getopt_long() doesn't care if the value comes in the same argv[] as the option or in the next one.  Pretty common pattern to do mashed-together option+val for short options.  All these are legal and mean the same thing:

-w2
-w 2
--wait 2

thockin commented:
Oh snap!  I suddenly hate our implementation of -w2.  Your assessment seems
OK, though.

On Thu, Mar 23, 2017 at 11:19 AM, Dan Williams <notifications@github.com>
wrote:

> iptables -w2 flag is not "wait 2 seconds" but "v2 of the wait flag". Is
> this new flag different?
>
> @thockin <https://github.com/thockin> no, it's the actual option value in
> seconds. getopt_long() doesn't care if the value comes in the same argv[]
> as the option or in the next one. Pretty common pattern to do
> mashed-together option+val for short options. All these are legal and mean
> the same thing:
>
> -w2
> -w 2
> --wait 2
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/43575#issuecomment-288815121>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVJYZhnAmvtTbvb_DwwepDcnRYfIIks5rorfFgaJpZM4MnEGZ>
> .
>

thockin commented:
OTOH, is there any downside to a longer timeout?

On Thu, Mar 23, 2017 at 11:22 AM, Tim Hockin <thockin@google.com> wrote:

> Oh snap!  I suddenly hate our implementation of -w2.  Your assessment
> seems OK, though.
>
> On Thu, Mar 23, 2017 at 11:19 AM, Dan Williams <notifications@github.com>
> wrote:
>
>> iptables -w2 flag is not "wait 2 seconds" but "v2 of the wait flag". Is
>> this new flag different?
>>
>> @thockin <https://github.com/thockin> no, it's the actual option value
>> in seconds. getopt_long() doesn't care if the value comes in the same
>> argv[] as the option or in the next one. Pretty common pattern to do
>> mashed-together option+val for short options. All these are legal and mean
>> the same thing:
>>
>> -w2
>> -w 2
>> --wait 2
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/43575#issuecomment-288815121>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AFVgVJYZhnAmvtTbvb_DwwepDcnRYfIIks5rorfFgaJpZM4MnEGZ>
>> .
>>
>
>

dcbw commented:
> OTOH, is there any downside to a longer timeout?

@thockin other than blocking syncProxyRules() and holding the kube-proxy internal lock for longer, not that I can think of.  But we already block under the lock if the flowcontrol.Throttle isn't ready to allow sync.  I guess that just blocks the apiserver connection and lets events queue up?

thockin commented:
That's certainly how it SHOULD work, I'm hard pressed to guarantee it right
now.

On Thu, Mar 23, 2017 at 11:27 AM, Dan Williams <notifications@github.com>
wrote:

> OTOH, is there any downside to a longer timeout?
>
> @thockin <https://github.com/thockin> other than blocking
> syncProxyRules() and holding the kube-proxy internal lock for longer, not
> that I can think of. But we already block under the lock if the
> flowcontrol.Throttle isn't ready to allow sync. I guess that just blocks
> the apiserver connection and lets events queue up?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/43575#issuecomment-288817323>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVEGBDFieyCfUtVuyGi0Y5Q5TGZbWks5rormCgaJpZM4MnEGZ>
> .
>

eparis commented:
thockin hates posix flags? I hate golang flags that `-bob` == `--bob`.  That's not how unix does things.  `-bob` is equivalent to `-b -o -b`

dcbw commented:
> -bob is equivalent to -b -o -b

@eparis or "-b ob" depending on how many : you give getopt(), what fun!

thockin commented:
I don't hate POSIX, I just made a bad assumption that -w2 was a different
flag, but of course, it's not if you assume POSIX.

On Thu, Mar 23, 2017 at 11:44 AM, Eric Paris <notifications@github.com>
wrote:

> thockin hates posix flags? I hate golang flags that -bob == --bob. That's
> not how unix does things. -bob is equivalent to -b -o -b
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/43575#issuecomment-288822465>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVFcmX4J3ntwMeyjGvgETGog3Gpbjks5ror2dgaJpZM4MnEGZ>
> .
>

danwinship commented on pkg/util/iptables/iptables.go:
```diff
@@ -503,6 +518,18 @@ func getIPTablesWaitFlag(vstring string) []string {
+	bytes, err := exec.Command(cmdIPTablesRestore, "--help").CombinedOutput()
+	if err != nil {
+		return restoreWaitUnknown
+	} else if strings.Contains(string(bytes), "[-w secs]") {
```
That seems pretty fragile...

It looks like iptables-restore doesn't even have a --version flag though anyway. And doesn't exit with an error if you pass invalid options. Sigh.

You could check if `iptables-restore -w < /dev/null` prints "invalid option"...

danwinship replied:
(well, ok, that's still pretty fragile... I'd initially written "if it's silent or prints something to stderr", but then I realized maybe something else could go wrong so that's not a good check either.)

Maybe the fix is to quickly get them to add a --version flag, and then if `iptables-restore --version` prints *anything*, you know it's good...

dcbw replied:
@danwinship done; submitted a patch to netfilter-devel@ to add --version/-V argument; if that gets accepted I will update this PR to check for version 1.6.2 and later.  Still doesn't help us for backported wait patches (like in RHEL) though, thoughts on that?  Would we just have to patch locally for OpenShift or should we still keep the help text checking here?

danwinship replied:
I meant we should just backport the --version patch to RHEL too, and then that's why I said "if it prints *anything*, you know it's good"; just assume that any version of iptables-restore that supports --version also supports -w, regardless of the actual version number.

(If that sounds too crazy to you/other reviewers then I guess check for 1.6.2 in kubernetes and we can patch it somehow in OpenShift.)

thockin replied:
I can live with this in the mean time, since you will be the primary consumers of this for a while, it would be great to have a "clean" upstream, but I won't argue if you want to stick with this.

dcbw replied:
@danwinship RHEL7 currently ships iptables 1.4.21 which we first built for RHEL on Jan 9 2014.  Since then we've backported about 6 patches from git.  I seriously doubt that RHEL will just rebase iptables to 1.6.2 whenever that deigns to come out.

What about a two-stage solution: first check version, then if that fails fall back to --help parsing?

dcbw replied:
Hmm, misread @danwinship's comment.  Yes, I can accept just checking for --version output and using that as a switch.  Let me see if we can get that backported for RHEL packages.

dcbw commented:
Ok, we now have --version in iptables-restore.  I'll update the PR to check for that.

dcbw commented:
@danwinship and then again, we can't use it.  iptables-restore ignores unknown arguments, and will sit there and wait for input if it find them.  So we do have to use the "--help" approach.

Thoughts?  Good to go as-is?

thockin commented:
ugh, ok.  Is it ready for review, then?

On Thu, Apr 13, 2017 at 8:49 PM, Dan Williams <notifications@github.com>
wrote:

> @danwinship <https://github.com/danwinship> and then again, we can't use
> it. iptables-restore ignores unknown arguments, and will sit there and wait
> for input if it find them. So we do have to use the "--help" approach.
>
> Thoughts? Good to go as-is?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/43575#issuecomment-294079450>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVHm2ZkOvISyKMz2QM08G3hIqXXhQks5rvuzYgaJpZM4MnEGZ>
> .
>

danwinship commented:
> iptables-restore ignores unknown arguments, and will sit there and wait for input if it find them.

Hm. You can do the equivalent of "iptables-restore --version < /dev/null" by calling SetStdin() on the kexec.Cmd before calling CombinedOutput() on it.

dcbw commented:
@danwinship another thought; can we assume if you have /usr/sbin/iptables v1.6.1 that you also have iptables-restore v1.6.1?  Or should we really check the version of each one separately?

dcbw commented:
@danwinship @thockin how about now?  I implemented Winship's suggestion to check for --version only and work around the non-exit option with Stdin().  Added a new testcase for old iptables-restore responses.

thockin commented:
You just tell me when you two are happy :)  The idea is fine with me, the
details are yours


On Wed, Apr 19, 2017 at 6:41 PM, Dan Williams <notifications@github.com>
wrote:

> @danwinship <https://github.com/danwinship> @thockin
> <https://github.com/thockin> how about now? I implemented Winship's
> suggestion to check for --version only and work around the non-exit option
> with Stdin(). Added a new testcase for old iptables-restore responses.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/43575#issuecomment-295535648>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVJvGQg5ZDI8X7NNtOV6HsW-lWGpcks5rxrfDgaJpZM4MnEGZ>
> .
>

danwinship commented on pkg/util/iptables/iptables.go:
```diff
@@ -519,6 +528,45 @@ func getIPTablesVersionString(exec utilexec.Interface) (string, error) {
+		return nil
+	}
+
+	return []string{"--wait", "10"}
```
The amount of time we might have to wait for the lock depends on whether the *other* operation is "iptables" or "iptables-restore", not on whether *this* operation is. So it doesn't make sense for this value to be different from the iptablesWaitFlag value; if 2 seconds isn't enough here then it's not enough there either. (And it's not enough in docker...)

dcbw replied:
back to 2

danwinship commented on pkg/util/iptables/iptables.go:
```diff
@@ -142,12 +143,19 @@ func New(exec utilexec.Interface, dbus utildbus.Interface, protocol Protocol) In
+
+	restoreVstring, err := getIPTablesRestoreVersionString(exec)
+	if err != nil {
+		glog.Warningf("Error checking iptables-restore version, assuming version at least %s: %v", MinCheckVersion, err)
```
the warning message is incorrect

danwinship replied:
also, in the case of "iptables", we do multiple things with the version string (and also call the get-version-string method from other code), but for iptables-restore we don't so it might be simpler to squash getIPTablesRestoreVersionString into getIPTablesRestoreWaitFlag?

dcbw replied:
done

danwinship reviewed (commented):
Basic functionality looks good now. Just some nitpicks.

danwinship commented:
> @danwinship another thought; can we assume if you have /usr/sbin/iptables v1.6.1 that you also have iptables-restore v1.6.1? Or should we really check the version of each one separately?

I'd be fine with assuming that, but that wouldn't help with the RHEL backport problem...

dcbw commented:
> I'd be fine with assuming that, but that wouldn't help with the RHEL backport problem...

@danwinship pretty much nothing is going to help us with that, since it appears we won't get that backport for a while anyway.  So we get to work around it in OpenShift.

dcbw commented:
@danwinship PTAL, thanks!

danwinship reviewed (approved):
LGTM

thockin commented:
/lgtm
/approve
